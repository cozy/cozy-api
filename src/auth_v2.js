/* global btoa */
const V2TOKEN_ABORT_TIMEOUT = 3000

export function getAppToken () {
  return new Promise(function (resolve, reject) {
    if (typeof window === 'undefined') {
      return reject(new Error('getV2Token should be used in browser'))
    } else if (!window.parent) {
      return reject(new Error('getV2Token should be used in iframe'))
    } else if (!window.parent.postMessage) {
      return reject(new Error('getV2Token should be used in modern browser'))
    }
    const origin = window.location.origin
    const intent = {action: 'getToken'}
    let timeout = null
    const receiver = function (event) {
      let token
      try {
        token = new AppToken({
          appName: event.data.appName,
          token: event.data.token
        })
      } catch (e) {
        reject(e)
        return
      }
      window.removeEventListener('message', receiver)
      clearTimeout(timeout)
      resolve({ client: null, token })
    }
    window.addEventListener('message', receiver, false)
    window.parent.postMessage(intent, origin)
    timeout = setTimeout(() => {
      reject(new Error('No response from parent iframe after 3s'))
    }, V2TOKEN_ABORT_TIMEOUT)
  })
}

export class AppToken {
  constructor (opts) {
    this.appName = opts.appName || ''
    this.token = opts.token || ''
  }

  toAuthHeader () {
    return 'Basic ' + btoa(`${this.appName}:${this.token}`)
  }
}
