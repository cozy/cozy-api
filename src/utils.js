/* global fetch, btoa */

export let config = {}

export function configure (opts) {
  config = opts
  return config
}

export async function waitConfig (opts) {
  return config
}

export function promiser (promise, optCallback) {
  if (!optCallback || typeof optCallback !== 'function') {
    return promise
  }
  promise.then(
    function (result) { optCallback(null, result) },
    function (err) { optCallback(err, null) }
  )
  return null
}

export function createPath (config, doctype, id = '', query = null) {
  let route = '/data/'
  if (!config.isV2) {
    route += `${encodeURIComponent(doctype)}/`
  }
  if (id !== '') {
    route += encodeURIComponent(id)
  }
  if (query) {
    const q = []
    for (const qname in query) {
      q.push(`${encodeURIComponent(qname)}=${encodeURIComponent(query[qname])}`)
    }
    route += `?${q.join('&')}`
  }
  return route
}

export function doFetch (config, method, path, body) {
  const options = {method: method, headers: {}}
  if (body !== undefined) {
    options.headers['Content-Type'] = 'application/json'
    options.body = JSON.stringify(body)
  }
  if (config.auth) {
    let auth = config.auth.appName + ':' + config.auth.token
    options.headers['Authorization'] = 'Basic ' + btoa(auth)
  }

  let target = config.target || ''
  let pathprefix = config.isV2 ? '/ds-api' : ''
  let fullpath = target + pathprefix + path
  return fetch(fullpath, options).then((res) => {
    const json = res.json()
    if (!res.ok) {
      return json.then(err => { throw err })
    } else {
      return json
    }
  })
}

const KNOWN_DOCTYPES = {
  'files': 'io.cozy.files',
  'folder': 'io.cozy.folders',
  'contact': 'io.cozy.contacts',
  'event': 'io.cozy.events',
  'track': 'io.cozy.labs.music.track',
  'playlist': 'io.cozy.labs.music.playlist'
}
const REVERSE_KNOWN = {}
Object.keys(KNOWN_DOCTYPES).forEach(k => { REVERSE_KNOWN[KNOWN_DOCTYPES[k]] = k })
export function normalizeDoctype (config, doctype) {
  let isQualified = doctype.indexOf('.') !== -1
  if (config.isV2 && isQualified) {
    let known = REVERSE_KNOWN[doctype]
    if (known) return known
    return doctype.replace(/\./g, '-')
  }
  if (config.isV3 && !isQualified) {
    let known = KNOWN_DOCTYPES[doctype]
    if (known) {
      warn('you are using a non-qualified doctype ' + doctype + ' assumed to be ' + known)
      return known
    }
    throw new Error('Doctype ' + doctype + ' should be qualified.')
  }
  return doctype
}

let warned = []
export function warn (text) {
  if (warned.indexOf(text) === -1) {
    warned.push(text)
    console.log('Warning', text)
  }
}
