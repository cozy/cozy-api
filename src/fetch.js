/* global fetch */
import {refreshToken, AccessToken} from './auth_v3'
import {retry} from './utils'
import jsonapi from './jsonapi'

export function cozyFetch (cozy, path, options = {}) {
  return cozy.fullpath(path).then((fullpath) => {
    let resp
    if (options.disableAuth) {
      resp = fetch(fullpath, options)
    } else if (options.manualAuthCredentials) {
      resp = cozyFetchWithAuth(cozy, fullpath, options, options.manualAuthCredentials)
    } else {
      resp = cozy.authorize().then((credentials) =>
        cozyFetchWithAuth(cozy, fullpath, options, credentials))
    }
    return resp.then(handleResponse)
  })
}

function cozyFetchWithAuth (cozy, fullpath, options, credentials) {
  if (credentials) {
    options.headers = options.headers || {}
    options.headers['Authorization'] = credentials.token.toAuthHeader()
  }

  // the option credentials:include tells fetch to include the cookies in the
  // request even for cross-origin requests
  options.credentials = 'include'

  return Promise.all([
    cozy.isV2(),
    fetch(fullpath, options)
  ]).then(([isV2, res]) => {
    if ((res.status !== 401 && res.status !== 400) || isV2 || !credentials) {
      return res
    }
    // we try to refresh the token only for OAuth, ie, the client defined
    // and the token is an instance of AccessToken.
    const { client, token } = credentials
    if (!client || !(token instanceof AccessToken)) {
      return res
    }
    return retry(() => refreshToken(cozy, client, token), 3)()
      .then((newToken) => cozy.saveCredentials(client, newToken))
      .then((credentials) => cozyFetchWithAuth(cozy, fullpath, options, credentials))
  })
}

export function cozyFetchJSON (cozy, method, path, body, options = {}) {
  options.method = method

  const headers = options.headers = options.headers || {}

  headers['Accept'] = 'application/json'

  if (method !== 'GET' && method !== 'HEAD' && body !== undefined) {
    if (headers['Content-Type']) {
      options.body = body
    } else {
      headers['Content-Type'] = 'application/json'
      options.body = JSON.stringify(body)
    }
  }

  return cozyFetch(cozy, path, options)
    .then(handleJSONResponse)
}

function handleResponse (res) {
  if (res.ok) {
    return res
  }
  let data
  const contentType = res.headers.get('content-type')
  if (contentType && contentType.indexOf('json') >= 0) {
    data = res.json()
  } else {
    data = res.text()
  }
  return data.then(err => {
    throw new FetchError(res, err)
  })
}

function handleJSONResponse (res) {
  const contentType = res.headers.get('content-type')
  if (!contentType || contentType.indexOf('json') < 0) {
    return res.text((data) => {
      throw new FetchError(res, new Error('Response is not JSON: ' + data))
    })
  }

  const json = res.json()
  if (contentType.indexOf('application/vnd.api+json') === 0) {
    return json.then(jsonapi)
  } else {
    return json
  }
}

export class FetchError extends Error {
  constructor (res, reason) {
    super()
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor)
    }
    // XXX We have to hardcode this because babel doesn't play nice when extending Error
    this.name = 'FetchError'
    this.response = res
    this.url = res.url
    this.status = res.status
    this.reason = reason

    Object.defineProperty(this, 'message', {
      value: reason.message ||
        (typeof reason === 'string' ? reason : JSON.stringify(reason))
    })
  }
}

FetchError.isUnauthorized = function (err) {
  // XXX We can't use err instanceof FetchError because of the caveats of babel
  return err.name === 'FetchError' && err.status === 401
}

FetchError.isNotFound = function (err) {
  // XXX We can't use err instanceof FetchError because of the caveats of babel
  return err.name === 'FetchError' && err.status === 404
}
