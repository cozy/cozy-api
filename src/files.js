/* global fetch Blob File */
import {waitConfig} from './utils'

const contentTypeOctetStream = 'application/octet-stream'

function doUpload (config, data, contentType, method, path) {
  if (!data) {
    throw new Error('missing data argument')
  }

  const headers = {}
  const options = { method, headers }

  // transform any ArrayBufferView to ArrayBuffer
  if (data.buffer && data.buffer instanceof ArrayBuffer) {
    data = data.buffer
  }

  const isBuffer = (typeof ArrayBuffer !== 'undefined' && data instanceof ArrayBuffer)
  const isFile = (typeof File !== 'undefined' && data instanceof File)
  const isBlob = (typeof Blob !== 'undefined' && data instanceof Blob)
  const isString = (typeof data === 'string')

  if (!isBuffer && !isFile && !isBlob && !isString) {
    throw new Error('invalid data type')
  }

  if (!contentType) {
    if (isBuffer) {
      contentType = contentTypeOctetStream
    } else if (isFile) {
      contentType = data.type || contentTypeOctetStream
    } else if (isBlob) {
      contentType = contentTypeOctetStream
    } else if (typeof data === 'string') {
      contentType = 'text/plain'
    }
  }

  headers['content-type'] = contentType

  const target = config.target || ''
  return fetch(`${target}${path}`, options)
    .then((res) => {
      const json = res.json()
      if (!res.ok) {
        return json.then(err => { throw err })
      } else {
        return json
      }
    })
}

export async function createFile (data, options) {
  const config = await waitConfig()
  if (config.isV2) {
    throw new Error('not implemented on v2')
  }

  let {name, folderId, contentType} = options || {}

  // handle case where data is a file and contains the name
  if (!name && typeof data.name === 'string') {
    name = data.name
  }

  if (typeof name !== 'string' || name === '') {
    throw new Error('missing name argument')
  }

  const path = `/files/${encodeURIComponent(folderId || '')}`
  const query = `?Name=${encodeURIComponent(name)}&Type=io.cozy.files`
  return doUpload(config, data, contentType, 'POST', `${path}${query}`)
}

export async function updateFile (data, options) {
  const config = await waitConfig()
  if (config.isV2) {
    throw new Error('not implemented on v2')
  }

  let {fileId, contentType} = options || {}

  if (typeof fileId !== 'string' || fileId === '') {
    throw new Error('missing fileId argument')
  }

  const path = `/files/${encodeURIComponent(fileId)}`
  return doUpload(config, data, contentType, 'PUT', path)
}
