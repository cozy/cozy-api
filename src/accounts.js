import {cozyFetchJSON} from './fetch'

export function create (cozy, type, login = '', password = '') {
  return cozyFetchJSON(cozy, 'POST', '/data/io.cozy.accounts', {
    account_type: type,
    status: 'PENDING',
    auth: {
      login: login,
      password: password
    }
  })
}
