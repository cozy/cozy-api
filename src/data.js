import { createPath } from './utils'
import { normalizeDoctype } from './doctypes'
import { cozyFetchJSON } from './fetch'

const NOREV = 'stack-v2-no-rev'

export function create(cozy, doctype, attributes) {
  return cozy.isV2().then(isV2 => {
    doctype = normalizeDoctype(cozy, isV2, doctype)
    if (isV2) {
      attributes.docType = doctype
    }
    const path = createPath(cozy, isV2, doctype, attributes._id)
    const httpVerb = attributes._id ? 'PUT' : 'POST'
    delete attributes._id
    return cozyFetchJSON(cozy, httpVerb, path, attributes).then(resp => {
      if (isV2) {
        return find(cozy, doctype, resp._id)
      } else {
        return resp.data
      }
    })
  })
}

export function find(cozy, doctype, id) {
  return cozy.isV2().then(isV2 => {
    doctype = normalizeDoctype(cozy, isV2, doctype)

    if (!id) {
      return Promise.reject(new Error('Missing id parameter'))
    }

    const path = createPath(cozy, isV2, doctype, id)
    return cozyFetchJSON(cozy, 'GET', path).then(resp => {
      if (isV2) {
        return Object.assign(resp, { _rev: NOREV })
      } else {
        return resp
      }
    })
  })
}

export function findMany(cozy, doctype, ids) {
  if (!(ids instanceof Array)) {
    return Promise.reject(new Error('Parameter ids must be a non-empty array'))
  }
  if (ids.length === 0) {
    // So users don't need to be defensive regarding the array content.
    // This should not hide issues in user code since the result will be an
    // empty object anyway.
    return Promise.resolve({})
  }

  return cozy.isV2().then(isV2 => {
    if (isV2) {
      return Promise.reject(new Error('findMany is not available on v2'))
    }

    const path = createPath(cozy, isV2, doctype, '_all_docs', {
      include_docs: true
    })

    return cozyFetchJSON(cozy, 'POST', path, { keys: ids })
      .then(resp => {
        const docs = {}

        for (const row of resp.rows) {
          const { key, doc, error } = row
          docs[key] = error ? { error } : { doc }
        }

        return docs
      })
      .catch(error => {
        if (error.status !== 404) return Promise.reject(error)

        // When no doc was ever created and the database does not exist yet,
        // the response will be a 404 error.
        const docs = {}

        for (const id of ids) {
          docs[id] = { error }
        }

        return docs
      })
  })
}

export function findAll(cozy, doctype, options = { include_docs: true }) {
  return cozy.isV2().then(isV2 => {
    if (isV2) {
      return Promise.reject(new Error('findAll is not available on v2'))
    }

    const path = createPath(cozy, isV2, doctype, '_all_docs', options)

    return cozyFetchJSON(cozy, 'POST', path, {})
      .then(resp => {
        const docs = []

        for (const row of resp.rows) {
          const { doc } = row
          // if not couchDB indexes
          if (!doc._id.match(/_design\//)) docs.push(doc)
        }
        return docs
      })
      .catch(error => {
        // the _all_docs endpoint returns a 404 error if no document with the given
        // doctype exists.
        if (error.status === 404) return []
        throw error
      })
  })
}

export function changesFeed(cozy, doctype, options) {
  return cozy.isV2().then(isV2 => {
    doctype = normalizeDoctype(cozy, isV2, doctype)
    const path = createPath(cozy, isV2, doctype, '_changes', options)
    return cozyFetchJSON(cozy, 'GET', path)
  })
}

export function update(cozy, doctype, doc, changes) {
  return cozy.isV2().then(isV2 => {
    doctype = normalizeDoctype(cozy, isV2, doctype)
    const { _id, _rev } = doc

    if (!_id) {
      return Promise.reject(new Error('Missing _id field in passed document'))
    }

    if (!isV2 && !_rev) {
      return Promise.reject(new Error('Missing _rev field in passed document'))
    }

    if (isV2) {
      changes = Object.assign({ _id }, changes)
    } else {
      changes = Object.assign({ _id, _rev }, changes)
    }

    const path = createPath(cozy, isV2, doctype, _id)
    return cozyFetchJSON(cozy, 'PUT', path, changes).then(resp => {
      if (isV2) {
        return find(cozy, doctype, _id)
      } else {
        return resp.data
      }
    })
  })
}

export function updateAttributes(cozy, doctype, _id, changes, tries = 3) {
  return cozy.isV2().then(isV2 => {
    doctype = normalizeDoctype(cozy, isV2, doctype)
    return find(cozy, doctype, _id)
      .then(doc => {
        return update(cozy, doctype, doc, Object.assign({ _id }, doc, changes))
      })
      .catch(err => {
        if (tries > 0) {
          return updateAttributes(cozy, doctype, _id, changes, tries - 1)
        } else {
          throw err
        }
      })
  })
}

export function _delete(cozy, doctype, doc) {
  return cozy.isV2().then(isV2 => {
    doctype = normalizeDoctype(cozy, isV2, doctype)
    const { _id, _rev } = doc

    if (!_id) {
      return Promise.reject(new Error('Missing _id field in passed document'))
    }

    if (!isV2 && !_rev) {
      return Promise.reject(new Error('Missing _rev field in passed document'))
    }

    const query = isV2 ? null : { rev: _rev }
    const path = createPath(cozy, isV2, doctype, _id, query)
    return cozyFetchJSON(cozy, 'DELETE', path).then(resp => {
      if (isV2) {
        return { id: _id, rev: NOREV }
      } else {
        return resp
      }
    })
  })
}
