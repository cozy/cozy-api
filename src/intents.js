import {cozyFetchJSON} from './fetch'

const intentClass = 'coz-intent'

// inject iframe for service in given element
function injectService (url, element, data) {
  const document = element.ownerDocument
  if (!document) throw new Error('Cannot retrieve document object from given element')

  const window = document.defaultView
  if (!window) throw new Error('Cannot retrieve window object from document')

  const iframe = document.createElement('iframe')
  iframe.setAttribute('src', decodeURI(url))
  iframe.classList.add(intentClass)
  element.appendChild(iframe)

  // Keeps only http://domain:port/
  const serviceOrigin = url.split('/', 3).join('/')

  return new Promise((resolve, reject) => {
    let handshaken = false
    const messageHandler = (event) => {
      if (event.origin !== serviceOrigin) return

      if (event.data === 'intent:ready') {
        handshaken = true
        return event.source.postMessage(data, event.origin)
      }

      window.removeEventListener('message', messageHandler)
      iframe.parentNode.removeChild(iframe)

      if (event.data === 'intent:error') {
        return reject(new Error('Intent error'))
      }

      return handshaken
        ? resolve(event.data)
        : reject(new Error('Unexpected handshake message from intent service'))
    }

    window.addEventListener('message', messageHandler)
  })
}

export function create (cozy, action, type, data = {}, permissions = []) {
  if (!action) throw new Error(`Misformed intent, "action" property must be provided`)
  if (!type) throw new Error(`Misformed intent, "type" property must be provided`)

  const createPromise = cozyFetchJSON(cozy, 'POST', '/intents', {
    data: {
      type: 'io.cozy.intents',
      attributes: {
        action: action,
        type: type,
        data: data,
        permissions: permissions
      }
    }
  })

  createPromise.start = (element) => {
    return createPromise.then(intent => {
      let service = intent.attributes.services && intent.attributes.services[0]

      if (!service) {
        return Promise.reject(new Error('Unable to find a service'))
      }

      return injectService(service.href, element, data)
    })
  }

  return createPromise
}

function listenClientData (intent, window) {
  return new Promise((resolve, reject) => {
    const messageEventListener = (event) => {
      if (event.origin !== intent.attributes.client) return

      window.removeEventListener('message', messageEventListener)
      resolve(event.data)
    }

    window.addEventListener('message', messageEventListener)
    window.parent.postMessage('intent:ready', intent.attributes.client)
  })
}

// returns a service to communicate with intent client
export function createService (cozy, id, window) {
  return cozyFetchJSON(cozy, 'GET', `/intents/${id}`)
    .then(intent => {
      return listenClientData(intent, window)
        .then(data => {
          let terminated = false
          return {
            getData: () => data,
            getIntent: () => intent,
            terminate: (doc) => {
              if (terminated) throw new Error('Intent service has already been terminated')
              terminated = true
              window.parent.postMessage(doc, intent.attributes.client)
            }
          }
        })
    })
}
