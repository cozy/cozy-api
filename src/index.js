import * as crud from './crud'
import * as mango from './mango'
import * as files from './files'
import init from './init'
import {promiser, warn} from './utils'

let cozy = {
  init: function (opts, optCallback) {
    return promiser(init(opts), optCallback)
  },
  // create(doctype, attributes) add a document to the database
  create: function (doctype, attributes, optCallback) {
    return promiser(crud.create(doctype, attributes), optCallback)
  },
  // find(doctype, id) retrieve a document by its doctype & ID.
  find: function (doctype, id, optCallback) {
    return promiser(crud.find(doctype, id), optCallback)
  },
  update: function (doctype, doc, changes, optCallback) {
    return promiser(crud.update(doctype, doc, changes), optCallback)
  },
  delete: function (doctype, doc, optCallback) {
    return promiser(crud._delete(doctype, doc), optCallback)
  },
  defineIndex: function (doctype, indexDef, optCallback) {
    return promiser(mango.defineIndex(doctype, indexDef), optCallback)
  },
  query: function (indexRef, query, optCallback) {
    return promiser(mango.query(indexRef, query), optCallback)
  },
  // updateAttributes(doctype, id, changes) performs a patch.
  updateAttribute: function (doctype, id, changes, optCallback) {
    return promiser(crud.updateAttributes(doctype, id, changes), optCallback)
  },
  // aliased delete to destroy fo browser-sdk compatibility
  destroy: function (doctype, doc, optCallback) {
    warn('destroy is deprecated, use cozy.delete instead.')
    return promiser(crud._delete(doctype, doc), optCallback)
  },
  createFile: function (data, options, optCallback) {
    return promiser(files.createFile(data, options), optCallback)
  },
  updateFile: function (data, options, optCallback) {
    return promiser(files.updateFile(data, options), optCallback)
  }
}

if ((typeof window) !== 'undefined') window.cozy = cozy

export default cozy
