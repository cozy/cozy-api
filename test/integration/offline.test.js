/* eslint-env jest */
/* global jasmine */

// eslint-disable-next-line no-unused-vars
import 'isomorphic-fetch'
import { Client } from '../../src'
import PouchDB from 'pouchdb'
import pouchdbFind from 'pouchdb-find'
import { sleep } from '../../src/utils'
PouchDB.plugin(require('pouchdb-adapter-memory'))

// PouchDB should not be a mandatory dependency as it is only used in mobile
// environment, so we declare it in global scope here.
global.PouchDB = PouchDB
global.pouchdbFind = pouchdbFind

const COZY_STACK_URL = process.env && process.env.COZY_STACK_URL || ''
const COZY_STACK_VERSION = process.env && process.env.COZY_STACK_VERSION
const COZY_STACK_TOKEN = process.env && process.env.COZY_STACK_TOKEN
const DOCTYPE = 'io.cozy.testobject2'

let docs = [
  {group: 'A', path: '/a/b/c', year: 1984},
  {group: 'A', path: '/a', year: 1749},
  {group: 'A', path: '/a/z', year: 2104},
  {group: 'A', path: '/a/e', year: 1345},
  {group: 'B', path: '/a/b/d', year: 2104}
]

describe('offline', function () {
  const cozy = {}

  beforeAll(async function () {
    if (COZY_STACK_VERSION === '2') {
      return this.skip()
    }
    cozy.client = new Client({
      cozyURL: COZY_STACK_URL,
      token: COZY_STACK_TOKEN
    })
    docs = await Promise.all(docs.map(doc => cozy.client.data.create(DOCTYPE, doc)))
  })

  afterAll(async function () {
    if (COZY_STACK_VERSION === '3') {
      await docs.forEach(doc => cozy.client.data.delete(DOCTYPE, doc))
      await cozy.client.offline.destroyAllDatabase()
    }
  })

  let originalTimeout

  beforeEach(function () {
    originalTimeout = jasmine.DEFAULT_TIMEOUT_INTERVAL
    jasmine.DEFAULT_TIMEOUT_INTERVAL = 10000
  })

  afterEach(function () {
    jasmine.DEFAULT_TIMEOUT_INTERVAL = originalTimeout
  })

  it('can replicate database from cozy', async function () {
    await cozy.client.offline.createDatabase(DOCTYPE, {adapter: 'memory'})
    let complete = await cozy.client.offline.replicateFromCozy(DOCTYPE)
    expect(complete.docs_written).not.toBe(0)
    complete = await cozy.client.offline.replicateFromCozy(DOCTYPE)
    return expect(complete.docs_written).toBe(0)
  })

  it('can\'t replicate with live option.', async function () {
    await cozy.client.offline.createDatabase(DOCTYPE, {adapter: 'memory'})
    return expect(cozy.client.offline.replicateFromCozy(DOCTYPE, {live: true})).rejects.toThrowErrorMatchingSnapshot()
  })

  it('can replicate created object in local database', async function () {
    // create a database
    const db = await cozy.client.offline.createDatabase(DOCTYPE, { adapter: 'memory' })
    // create a doc
    const sampleDoc = { data: 'some Data' }
    const remoteDoc = await cozy.client.data.create(DOCTYPE, sampleDoc)
    // check the db to look for the new doc
    await expect(db.get(remoteDoc._id)).rejects.toThrowErrorMatchingSnapshot()
    // replicate the database
    await cozy.client.offline.replicateFromCozy(DOCTYPE)
    // doc should exist
    await expect(db.get(remoteDoc._id)).resolves.toEqual({ _id: remoteDoc._id, _rev: remoteDoc._rev, data: remoteDoc.data })
    // remove doc
    return cozy.client.data.delete(DOCTYPE, remoteDoc)
  })

  it.skip('can repeated replication and stop repeated replication', async function () {
    // create a database
    const db = await cozy.client.offline.createDatabase(DOCTYPE, { adapter: 'memory' })
    // create a doc
    const sampleDoc = { data: 'some Data' }
    const remoteDoc = await cozy.client.data.create(DOCTYPE, sampleDoc)
    // check the db to look for the new doc
    expect(db.get(remoteDoc._id)).rejects.toThrowErrorMatchingSnapshot()
    // activate synchronisation x ms
    expect(cozy.client.offline.hasRepeatedReplication(DOCTYPE)).toBe(false)
    cozy.client.offline.startRepeatedReplication(DOCTYPE, 0.5)
    // after a certain amount of time, doc should exist
    await sleep(1000)
    expect(cozy.client.offline.hasRepeatedReplication(DOCTYPE)).toBe(true)
    const resp = await cozy.client.offline.stopRepeatedReplication()
    expect(resp).toEqual({ ok: true })
    expect(cozy.client.offline.hasRepeatedReplication(DOCTYPE)).toBe(false)
    // create another doc after sync
    const anotherDoc = { data: 'some other Data' }
    const anotherRemoteDoc = await cozy.client.data.create(DOCTYPE, anotherDoc)
    const promises = []
    promises.push(expect(db.get(remoteDoc._id)).resolves.toEqual({ _id: remoteDoc._id, _rev: remoteDoc._rev, data: remoteDoc.data }))
    promises.push(expect(db.get(anotherRemoteDoc._id)).rejects.toThrowErrorMatchingSnapshot())
    // remove docs
    cozy.client.data.delete(DOCTYPE, remoteDoc)
    cozy.client.data.delete(DOCTYPE, anotherRemoteDoc)
    return Promise.all(promises)
  })
})
