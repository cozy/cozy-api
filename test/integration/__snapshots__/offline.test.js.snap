// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`offline only v3 can replicate created object in local database 1`] = `"received is not a function"`;

exports[`offline only v3 can't replicate with live option. 1`] = `"received is not a function"`;
